### PR TITLE
#64 - Handle partially failable responses

### PR DESCRIPTION
**Pull request by alessandrodalbello**
This is to fix how our HTTP callback implementation handles server 400 responses that contain partially failed requested data, such as multiple offers submission requests which include some offers correctly placed and at least one failed offer, perhaps because the runner has been closed before the request reached the server back-end.

That's a detailed summary of the changes to help you review the code:

- new interfaces created for REST requests that might fail. In particular, the scenario we want to handle can happen only in `PageableResponse`s and only for some kind of `RestResponse`s. Hence, I defined a new `FailbaleRestResponse` and extended the `PageableResponse` class in a `PartiallyFailableResponse` including `FailbaleRestResponse`s and which in turns is a `FailbaleRestResponse` itself (composite pattern);
- new interfaces applied to the following REST responses:

    - `Offer` to implement `FailbaleRestResponse` (instead of `RestResponse`)
    - `OfferEdit` to implement `FailbaleRestResponse` (instead of `RestResponse`)
    - `OffersResponse` to extend `PartiallyFailableResponse` (instead of `PageableResponse`)
    - `OfferEditsResponse` to extend `PartiallyFailableResponse` (instead of `PageableResponse`)
- adjusted the `HttpCallback` interface to handle the new response scenarios and extracted the existing `RestResponseCallback` inner class in a dedicated class that behaves as before from a higher perspective, but with some logic refactored. Of course, still this class is not aware of the actual library used by the `HttpClient` wrapper;
- extended the `RestResponseCallback` in a `PartiallyFailableResponseCallback` that acts as normal when the server response is successful, though overwrites the way failed responses are handled;
- adjusted the base REST client implementation in order to instantiate either  `RestResponseCallback`s or `PartiallyFailableResponseCallback`s based on the request. Please note even if a response is failable it doesn't mean it can fail on every request. For instance, `OffersResponse` is received in GET, POST, PUT and DELETE requests, but it cannot fail for GET requests. Thus, the implementation of the `OffersClientRest.getOffers()` method does not implement a `PartiallyFailableResponseCallback` to handle server replies;
- during refactoring of `HttpCallback`s I realised the existing readers were not working well with empty responses (an edge case should never occur). Empty responses were handled in a naive way and only in non-200 responses. I fixed that in the abstract `Reader` used by all concreate implementations;
- finally, integration tests have been refactored to extract two specific `StreamObserver`s implementations to verify the behaviour of successful and failed responses.

### CODE REVIEW

File: matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        if (responseCode == FAILABLE_HTTP_CODE) {
+            onFailableResponse(responseInputStream, responseCode);
+        } else {
+            super.onFailedResponse(responseInputStream, responseCode);
+        }
+    }
+
+    private void onFailableResponse(InputStream responseInputStream, int responseCode) {
+        try (ByteArrayOutputStream responseBytes = copyBytes(responseInputStream);
+                InputStream copiedInputStream = copyInputStream(responseBytes)) {
```
Comment (volkodavs) [inline matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java]: why do you copy the input stream?

File: matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            try (Parser parser = serializer.newParser(copiedInputStream)) {
+                boolean isPartiallyFailedResponse = tryPartiallyFailedResponse(parser);
+                if (isPartiallyFailedResponse) {
+                    observer.onCompleted();
+                } else {
+                    secondParsing(responseBytes, responseCode);
+                }
```
Comment (volkodavs) [inline matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java]: can be an error in the middle of response?

File: matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/dtos/FailableRestResponse.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import com.matchbook.sdk.rest.dtos.errors.Errors;
+
+public interface FailableRestResponse extends RestResponse {
+
+    Errors getErrors();
+
+    default boolean hasFailed() {
+        Errors errors = getErrors();
+        return Objects.nonNull(errors) && !errors.getErrors().isEmpty();
+    }
```
Comment (volkodavs) [inline matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/dtos/FailableRestResponse.java]: :cat:

File: matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java (hunk 1 above)
Comment (alessandrodalbello) [reply matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java]: Unfortunately, the only way to know if the response contains a partially failed content or a full error message is to start parsing the bytes. If I parse the input stream, I'll consume it and make it unavailable. The only solution I know in order to be capable to parse the content twice is to copy the bytes is a swapping data structure, parse it once and only if necessary parse the copy another time.

Of course this way we can not take advantage of the stream parsers. However, we use this technique only for responses that might partially fail and only when the response HTTP code suggests it could be the case.

File: matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java (hunk 2 above)
Comment (alessandrodalbello) [reply matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java]: Yes, the process will fail if the content is not the `PartiallyFailableResponse` we try to parse in the first attempt. Specifically, the first parsing attempt could either throw an exception or not be able to extract any meaningful information from the content. In both these cases, a second parse will be invoked to read a standard error content.

Honestly, I don't really the methods name I used (i.e. `tryPartiallyFailedResponse`, `secondParsing`). Any suggestion?

File: matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java (hunk 1 above)
Comment (volkodavs) [reply matchbook-sdk-rest/src/main/java/com/matchbook/sdk/rest/PartiallyFailableResponseCallback.java]: yeah, i got you. Let's leave it like that at the moment, but we should improve this in a future
